### PR TITLE
Add --duplicated-nevra "keep-last" option, and --delayed-dump

### PR DESCRIPTION
**Pull request by praiskup**
This is follow-up for #307.  Unfortunately, the parallel processing performance degrades a bit
with the "keep-last" option.

Motivation for Copr: https://pagure.io/copr/copr/issue/840

We can tweak this in the future (keep latest pkg per mtime, keep the first one, etc.).

### CODE REVIEW

**Comment by praiskup** [comment]
> parallel processing performance degrades a bit

By that I mean:
- using `createrepo_c --no-database --update --skip-stat --recycle-pkglist --pkglist add .  --workers 8`
- on a repository with 250k+ packages, and only one is being added, i.e. mostly a CPU bound task (not I/O)
- the slow-down is from 55s to 91s on my machine

We could start another pool of workers there, but I'm not sure it is needed at this point in time.

**Comment by dralley** [comment]
Presumably in a situation where there are many duplicates, this would significantly improve the metadata size correct?

**Comment by praiskup** [comment]
> Presumably in a situation where there are many duplicates, this would
significantly improve the metadata size correct?

Yes.  The delayed-write logic can also be tweaked to implement any
additional filtering mechanism in the future (if metadata size is a
concern).

**Comment by dralley** [comment]
`keep-last` should probably keep the most recent package as determined by a second-order sort by build time, which is what yum / dnf / zypper does when encountering multiple packages of the same NEVRA.

**Review by kontura — CHANGES_REQUESTED** [review]
> `keep-last` should probably keep the most recent package as determined by a second-order sort by build time, which is what yum / dnf / zypper does when encountering multiple packages of the same NEVRA.

I would also like the build-time ordering. It would break some current CI tests but I can easily fixed that.

File: src/createrepo_c.c [hunk 1]
```diff
@@ -1416,7 +1462,7 @@ main(int argc, char **argv)
      * that unfortunately means we have to decompress metadata files change package
      * count value and compress them again.
      */
-    if (user_data.package_count != user_data.task_count){
+    if (user_data.package_count != user_data.task_count - skipped_pkgs){
```
Comment (kontura) [inline src/createrepo_c.c:1465]: The `user_data.task_count` is also used in the [`cr_rewrite_pkg_count_thread()`](https://github.com/rpm-software-management/createrepo_c/blob/master/src/threads.c#L134) which causes problems because the `skipped_pkgs` are not taken into account. 

It can be reproduced by running with invalid packages (such as `echo "test" > bad.rpm`).

**Comment by praiskup** [comment]
Thank you for the review!  Updated but WIP, I still need to implement build-time ordering.

File: src/dumper_thread.c [hunk 2]
```diff
@@ -233,6 +233,54 @@ write_pkg(long id,
...
+void
+cr_delayed_dump_run(gpointer user_data)
+{
+    GError *tmp_err = NULL;
+    struct UserData *udata = (struct UserData *) user_data;
+    long int stop = udata->task_count;
+    g_debug("Performing the delayed metadata dump");
+    for (int id = 0; id < stop; id++) {
+        struct DelayedTask dtask = g_array_index(udata->delayed_write,
+                                                 struct DelayedTask, id);
```
[the remark this answers is not in the document]
Comment (praiskup) [reply src/dumper_thread.c:262]: Good point, all IDs are now properly initialized in the delayed list.

**Comment by praiskup** [comment]
> keep-last should probably keep the most recent package as determined by a second-order sort by build time, which is what yum / dnf / zypper does when encountering multiple packages of the same NEVRA.

So, trying to understand what actually DNF does, and it appears it prefers the oldest package from all with the NVR, aka the oldest per build time.  Is it a correct observation?

**Comment by dralley** [comment]
AFAIK it prefers the newest per NEVRA, and the newest per build-time between packages with the same NEVRA.

**Comment by kontura** [comment]
For dnf I think this depends on the order of packages in metadata. I believe libsolv picks just the **first** package in the metadata.
I made a hacky CI test to demonstrate: https://github.com/rpm-software-management/ci-dnf-stack/pull/1173

However this is just for `dnf install package-name` we might be doing something else for different arguments/commands. 
It is possible that different filtering will send different `id`s to libsolv resulting in installing the other package but I don't think there is any dedicated logic to it. It might happen by accident if at all.

**Comment by praiskup** [comment]
Still experimenting a bit, and DNF seems to prefer the one which comes first
sorted alphabetically (or that comes first in the repodata, because createrepo_c
sorts the packages alphabetically).

For example [this repo](https://copr-be.cloud.fedoraproject.org/archive/issues/createrepo_c-pr-325/repo2/), in a Rawhide container:
```
# dnf -y install dummy-pkg --repofrompath=ordering,https://copr-be.cloud.fedoraproject.org/archive/issues/createrepo_c-pr-325/repo1/ --setopt ordering.gpgcheck=0
...
Installed:
  dummy-pkg-20221031_1057-1.fc38.x86_64 
...
# tail /var/log/dnf.librepo.log | grep dummy-pkg
https://copr-be.cloud.fedoraproject.org/archive/issues/createrepo_c-pr-325/repo2/a/dummy-pkg-20221031_1057-1.fc38.x86_64.rpm
```

The one starting with `../a/dummy-pkg...` was installed.  But when you look at RPM headers:

```
# for rpm in `find -name '*x86_64.rpm'`; do echo "== $rpm =="; rpm -qpi "$rpm" 2>/dev/null| grep Build\ Date; done
== ./a/dummy-pkg-20221031_1057-1.fc38.x86_64.rpm ==
Build Date  : Mon 31 Oct 2022 10:00:35 AM UTC
== ./b/dummy-pkg-20221031_1057-1.fc38.x86_64.rpm ==
Build Date  : Mon 31 Oct 2022 10:00:35 AM UTC
== ./c/dummy-pkg-20221031_1057-1.fc38.x86_64.rpm ==
Build Date  : Tue 01 Nov 2022 08:59:25 AM UTC
== ./d/dummy-pkg-20221031_1057-1.fc38.x86_64.rpm ==
Build Date  : Mon 31 Oct 2022 10:00:21 AM UTC
```

Then the newest one is `./c` which was built today.  Though I agree that what
you describe @dralley is the behavior I'd expect and prefer.  @j-mracek is this a bug,
accident, or expected behavior? :-)

**Comment by praiskup** [comment]
@kontura was faster! :-)  Good to have this tested.

Anyway, for the `keep-last` option; should we do a two-level sort, by build time (which might still have collisions) and then by path name alphabetically?

**Comment by kontura** [comment]
> Anyway, for the `keep-last` option; should we do a two-level sort, by build time (which might still have collisions) and then by path name alphabetically?
 
That sounds reasonable to me. :slightly_smiling_face:

**Comment by praiskup** [comment]
Ok, please take another look.

File: src/createrepo_c.c [hunk 3]
```diff
@@ -1340,30 +1415,45 @@ main(int argc, char **argv)
...
     gpointer key, value;
 
+    int skipped_pkgs = 0;
     g_hash_table_iter_init(&iter, user_data.nevra_table);
     while (g_hash_table_iter_next(&iter, &key, &value))
     {
         gchar *nevra = (gchar *) key;
         GArray *locations = (GArray *) value;
         if (locations->len > 1) {
-            g_warning("Package '%s' has duplicate metadata entries, only one should exist", nevra);
```
Comment (dralley) [inline src/createrepo_c.c:1349]: We should keep the warning in place, A) because not all users will use the option, and B) because even users who do use the option should probably be aware of potential issues even if we're applying our own heuristics to deal with it.

**Comment by dralley** [comment]
Are we sure libsolv has any deliberate strategy here at all, as opposed to some kind of parsing bug where various package entries attempt to overwrite each other?

Is there an easy way that we can ask libsolv "load this repo and then dump the solvables" to compare the input and what libsolv sees?

e.g. If some package was present twice in the metadata, would you see only one solvable, and would potentially have doubled up filelist and changelog metadata

[This comment makes me suspicious that this will be the case](https://github.com/openSUSE/libsolv/blob/master/ext/repo_rpmmd.c#L654-L668)

**Comment by kontura** [comment]
> Are we sure libsolv has any deliberate strategy here, as opposed to some kind of parsing bug where various package entries attempt to overwrite each other?

Oh I meant during solving. Libsolv does parse all of the packages and each has a unique id. If you do `install foo` libdnf basically finds all ids that match name `foo` and sends them to libsolv. Here the solver has to pick something, I would guess that from duplicate packages it takes the lowest id (that is the first duplicate in metadata).

> 
> Is there an easy way that we can ask libsolv "load this repo and then dump the solvables" to compare the input and what libsolv sees at the pool level (pre- any kind of actual dependency solving / choosing logic)?

There is exactly such a thing. Libsolv has these [tools](https://github.com/openSUSE/libsolv/tree/master/tools) (rpm libsolv-tools-0.7.22-3.fc37.x86_64). You can do:
```
$ repo2solv ./path/to/repo | dumpsolv
```

File: src/dumper_thread.c [hunk 4]
```diff
@@ -43,7 +43,7 @@ struct BufferedTask {
     long id;                        // ID of the task
     struct cr_XmlStruct res;        // XML for primary, filelists and other
     cr_Package *pkg;                // Package structure
-    int pkg_from_md;                // If true - package structure if from
+    gboolean clean;                 // If false - package structure if from
```
Comment (kontura) [inline src/dumper_thread.c:46]: This comment now has twice `If false..`. I suppose the second should be changed to `If true..`?

Also on the first line there is `If false - package structure if from` I think that should have been `If false - package structure is from`. It would be nice if we could fix that as well.

**Comment by dralley** [comment]
>Oh I meant during solving. Libsolv does parse all of the packages and each has a unique id (even duplicates).

I think that would work while parsing primary.xml since you can create new solvables, I'm just curious what happens if you encounter duplicate pkgid while parsing other.xml and filelists.xml.  The files and changelogs have to be associated back to solvables, and all it has to go off of is NEVRA and pkgid, so how would it pick which solvable the files and changelogs get associated with?  If they all get dumped onto one solvable, then you'd have duplicate solvables but with broken file deps and changelogs on some or all of them.

Basically similar to https://github.com/rpm-software-management/createrepo_c/issues/306

However I've just done some testing and nothing stood out as being obviously wrong, so maybe it has some magic up its sleeve.

File: src/createrepo_c.c [hunk 5]
```diff
@@ -578,6 +578,66 @@ static int strlensort(gconstpointer a, gconstpointer b)
...
+    // and then alphabetically by the rpm location
+    return g_strcmp0(a_loc->location, b_loc->location);
+}
+
+
+static int
+handle_nevra_duplicates(GArray *locations, CmdDupNevra option)
+{
+    int skipped = 0;
+    for (int i=0; i<locations->len; i++) {
```
Comment (kontura) [inline src/createrepo_c.c:607]: Here we compare signed `int i` with unsigned `locations->len`,
I know its already present elsewhere but these things show up in covscans and other analyzers.
Can you change the `i` to unsigned? To `size_t` perhaps?

File: src/createrepo_c.c [hunk 6]
```diff
@@ -578,6 +578,66 @@ static int strlensort(gconstpointer a, gconstpointer b)
...
+  g_warning("Package '%s' has duplicate metadata entries, only one should exist", nevra);
+
+  char *skip_reason= "";
+  if (option == CR_ARG_DUP_NEVRA_KEEP_LAST) {
+      skip_reason = " (not dumped, 'keep-last')";
+  }
+  for (int i=0; i<locations->len; i++) {
+      struct DuplicateLocation location = g_array_index(locations, struct
+                                                        DuplicateLocation, i);
+      g_warning("    Sourced from location: \'%s\', build timestamp: %d%s",
```
Comment (kontura) [inline src/createrepo_c.c:632]: The `time_build` is `gint64` so I think the format specified should be `%ld`.

File: src/cmd_parser.c [hunk 7]
```diff
@@ -213,6 +213,11 @@ static GOptionEntry cmd_entries[] =
       "Read the list of packages from old metadata directory and re-use it.  This "
       "option is only useful with --update (complements --pkglist and friends).",
       NULL },
+    { "delayed-dump", 0, G_OPTION_FLAG_HIDDEN, G_OPTION_ARG_NONE,
```
Comment (kontura) [inline src/cmd_parser.c:235]: Is there any benefit to having this as a hidden option? Maybe for debugging? But personally I wouldn't add it here at all.

Comment (praiskup) [reply src/cmd_parser.c:235]: I used it several times for debugging, yes.  There shouldn't be cons in having the option, so I'd prefer to keeping it?  Just please re-consider, and I'll drop if you really want.

**Comment by kontura** [comment]
> Basically similar to #306
> 
> However I've just done some testing and nothing stood out as being obviously wrong, so maybe it has some magic up its sleeve.

I was interested in how it is done so I spent some time digging into the libsolv parser and there is no magic. It has the same bug as you mentioned.
Its just that I didn't realize that `repo2solv` by default uses just primary. Testing it with options to use full filelists and changelogs:
```
repo2solv . -F -C | dumpsolv
```
It does add all the duplicate entries to the first `id`, duplicating them.

I am thinking what does this mean for dnf or other tools. dnf would show duplicated changelogs but the filelists probably just add some overhead?

**Comment by praiskup** [comment]
PR updated.

File: src/cmd_parser.c (hunk 7 above)
Comment (praiskup) [reply src/cmd_parser.c]: ACK, dropped the command line option then, thanks for checking. I kept the "internal" delayed_dump option just to not overcomplicate the commit changes if that is OK?

Comment (kontura) [reply src/cmd_parser.c]: Great, thanks!

**Review by kontura — APPROVED** [review]

**Comment by dralley** [comment]
@kontura So Pulp has some logic to attempt to pick the package that DNF would pick when syncing a repo that happens to contain duplicate NEVRA (it skips over the other duplicates), and currently it uses build time as the differentiator because that's what we thought it does.

Should we change that to instead just pick the first one?

**Comment by kontura** [comment]
I have tried syncing with DNF `reposync` and that keeps all the duplicates.
Just to make sure I understand correctly you mean when Pulp is doing the syncing right?

If they have duplicate NEVRA and pkgid (checksum) then libsolv picks the first one.

However for just duplicate NEVRA DNF works with all of them. For example when you try to install something both of them provide then yes it (libsolv) picks the first one. But if you do `install /path/to/file` and the file is only present in the second one - it uses the second one. 
I think that in general libsolv will use any package of the duplicates that satisfies the dependency resolution and since the packages are different it could be any of them.

Based on that if you want to match what DNF does you need all of them.

**Comment by dralley** [comment]
> Just to make sure I understand correctly you mean when Pulp is doing the syncing right?

Yeah.  The trouble is that Pulp encoded some optimistic assumptions early on that NEVRA would be unique, and it led to bugs when that wasn't the case, but we can't reject those repos because there's too many of them and it would be a pain in the ass, so instead we're just trying to pick the one that DNF / Yum / zypper would prefer.

In theory the dependencies ought to be the same between the packages unless the packager has been doing grievously stupid things anyway.
